### PR TITLE
Module #49: Security Applications

### PR DESCRIPTION
**Pull request by YuliiaLysak**

### CODE REVIEW

File: module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/model/user/UserPrincipal.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        if (authGroups == null) {
+            return Set.of();
+        }
+        Set<SimpleGrantedAuthority> grantedAuthorities = new HashSet<>();
+        authGroups.forEach(group ->
+                grantedAuthorities.add(new SimpleGrantedAuthority(group.getAuthGroup()))
+        );
+        return grantedAuthorities;
```
Comment (arubankov-clgx) [inline module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/model/user/UserPrincipal.java:32]: There is a better way with a Stream API:
```
.stream()
.map()
.collect(toSet());
```

Using `.forEach()` for collecting data is not a good practice.

File: module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/repository/user/UserRepository.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import javax.transaction.Transactional;
+import java.util.List;
+
+public interface UserRepository extends CrudRepository<User, Long> {
+    User findByEmail(String email);
+
+    @Query("UPDATE User u SET u.failedAttempt = ?1 WHERE u.email = ?2")
+    @Modifying
+    @Transactional
+    void updateFailedAttempts(int failAttempts, String email);
```
Comment (arubankov-clgx) [inline module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/repository/user/UserRepository.java:17]: [minor] Please put a `@Query` annotation closest to method name (for a better view).

File: module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/repository/user/UserRepository.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import edu.lysak.recipes.model.user.User;
+import org.springframework.data.jpa.repository.Modifying;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.CrudRepository;
+
+import javax.transaction.Transactional;
+import java.util.List;
+
+public interface UserRepository extends CrudRepository<User, Long> {
+    User findByEmail(String email);
```
Comment (arubankov-clgx) [inline module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/repository/user/UserRepository.java:12]: I would recommend to prefer `@Query` annotation for any custom repository method.

File: module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/service/user/UserService.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+import java.util.Date;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Service
+public class UserService implements UserDetailsService {
+    public static final int MAX_FAILED_ATTEMPTS = 3;
```
Comment (arubankov-clgx) [inline module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/service/user/UserService.java:20]: Why this constant is public?

File: module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/service/user/UserService.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    private static final long LOCK_TIME_DURATION = 5 * 60 * 1000; // 5 minutes
```
Comment (arubankov-clgx) [inline module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/service/user/UserService.java:21]: Would be great to extract this parameter to configuration.
You can use `@Value(${param-name:default-value})` though constructor.

File: module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/service/user/UserService.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+        return false;
+    }
+
+    public User getByEmail(String email) {
+        return userRepository.findByEmail(email);
+    }
+
+    public List<BlockedUser> getBlockedUsers() {
+        List<User> blockedUsers = userRepository.findAllBlockedUsers();
```
Comment (arubankov-clgx) [inline module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/service/user/UserService.java:80]: A rule of thumb: always use limitation for selecting data from DB.
You can hardcode this limitation (or better way - extract to config), but limitation should be.

File: module-28-spring-boot/task-5/src/test/java/edu/lysak/recipes/IntegrationTest.java [hunk 7]
```diff
@@ -23,6 +23,8 @@
 @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
 public class IntegrationTest {
 
+    // TODO: 06.04.2022 fix tests after adding spring security
```
Comment (arubankov-clgx) [inline module-28-spring-boot/task-5/src/test/java/edu/lysak/recipes/IntegrationTest.java:26]: Did you fix it?

File: module-28-spring-boot/task-5/src/test/java/edu/lysak/recipes/controller/RecipeControllerTest.java [hunk 8]
```diff
@@ -33,6 +33,8 @@
 @WebMvcTest(RecipeController.class)
 class RecipeControllerTest {
 
+    // TODO: 06.04.2022 fix tests after adding spring security
```
Comment (arubankov-clgx) [inline module-28-spring-boot/task-5/src/test/java/edu/lysak/recipes/controller/RecipeControllerTest.java:36]: Did you fix it?

**Review by arubankov-clgx — COMMENTED** [review]
Please review comments

File: module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/controller/AdminController.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    private final UserService userService;
+
+    public AdminController(UserService userService) {
+        this.userService = userService;
+    }
+
+    @GetMapping("/blocked-users")
+    public ResponseEntity<List<BlockedUser>> getAllBlockedUsers() {
+        List<BlockedUser> users = userService.getBlockedUsers();
+        return ResponseEntity.ok(users);
```
[the remark this answers is not in the document]
Comment (YuliiaLysak) [reply module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/controller/AdminController.java:22]: Done

File: module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/controller/AdminController.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+@RestController
+public class AdminController {
+    private final UserService userService;
+
+    public AdminController(UserService userService) {
+        this.userService = userService;
+    }
+
+    @GetMapping("/blocked-users")
```
[the remark this answers is not in the document]
Comment (YuliiaLysak) [reply module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/controller/AdminController.java:19]: Done

File: module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/model/user/AuthGroup.java [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+@Table(name = "auth_user_group")
+public class AuthGroup {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    private String email;
+
+    @Column(name = "auth_group")
```
[the remark this answers is not in the document]
Comment (YuliiaLysak) [reply module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/model/user/AuthGroup.java:22]: 👍

File: module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/model/user/UserPrincipal.java (hunk 1 above)
Comment (YuliiaLysak) [reply module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/model/user/UserPrincipal.java:32]: Good point, done

File: module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/service/user/UserService.java (hunk 4 above)
Comment (YuliiaLysak) [reply module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/service/user/UserService.java:20]: mistake, fixed

File: module-28-spring-boot/task-5/src/test/java/edu/lysak/recipes/IntegrationTest.java (hunk 7 above)
Comment (YuliiaLysak) [reply module-28-spring-boot/task-5/src/test/java/edu/lysak/recipes/IntegrationTest.java:26]: I will finish it later

File: module-28-spring-boot/task-5/src/test/java/edu/lysak/recipes/controller/RecipeControllerTest.java (hunk 8 above)
Comment (YuliiaLysak) [reply module-28-spring-boot/task-5/src/test/java/edu/lysak/recipes/controller/RecipeControllerTest.java:36]: I will finish it later

File: module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/config/CustomLoginFailureHandler.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+        if (user != null) {
+            if (user.isAccountNonLocked()) {
+                if (user.getFailedAttempt() < maxFailedAttempts - 1) {
+                    userService.increaseFailedAttempts(user);
+                } else {
+                    userService.lock(user);
+                    exception = new LockedException(
+                            String.format("Your account has been locked due to %s failed attempts." +
+                            " It will be unlocked after %s minutes.", maxFailedAttempts, (lockTimeDuration / 1000 / 60))
```
Comment (arubankov-clgx) [inline module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/config/CustomLoginFailureHandler.java:51]: To convert to some time unit, you can use `TimeUnit`:
`TimeUnit.MILLISECONDS.toSeconds(4500)`

**Review by arubankov-clgx — APPROVED** [review]

File: module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/config/CustomLoginFailureHandler.java (hunk 12 above)
Comment (YuliiaLysak) [reply module-28-spring-boot/task-5/src/main/java/edu/lysak/recipes/config/CustomLoginFailureHandler.java]: Thanks, great idea

**Pull request merged**